### PR TITLE
Partially revert #8042 and add CI in order to allow Python 3.12 support.

### LINKED ISSUE #8044: Add python 3.12 support
https://github.com/spacetelescope/jwst/pull/8042 noted that jwst does not currently support python 3.12. This is in part due to `scipy.signal.medfilt` limiting the `scipy` version.

One the scipy upper pin can be removed, re-add testing and support for python 3.12.

### PR DESCRIPTION
**Pull request by WilliamJamieson**
Resolves [JP-nnnn](https://jira.stsci.edu/browse/JP-nnnn)

Closes #8044

This PR enables Python 3.12 support for JWST, the main issue preventing this was solved by #8033.

**Checklist for maintainers**
- [ ] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [ ] updated relevant documentation
- [ ] added relevant milestone
- [x] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by WilliamJamieson** [comment]
Looks like jwst still uses `pkg_resources`. I'll open a separate PR fixing that problem. The devdeps failure looks real, but due to numpy 2.0 problems.

**Comment by WilliamJamieson** [comment]
This requires #8095 in order to function properly as `pkg_resources` has been removed by Python 3.12.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -46,7 +46,7 @@ jobs:
   check:
     uses: OpenAstronomy/github-actions-workflows/.github/workflows/tox.yml@v1
     with:
-      default_python: "3.11"
+      default_python: "3.12"
```
[the remark this answers is not in the document]
Comment (WilliamJamieson) [reply .github/workflows/ci.yml:49]: `default_python` is the version of Python the workflow installs by default for all the CI it is running. By default (if you don't state it) it is the "py3/Python 3" designated version of python by GitHub. That was recently updated to be Python 3.12

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -70,10 +70,11 @@ jobs:
...
         - linux: py310-xdist
         - linux: py311-xdist
-        - macos: py311-xdist
+        - linux: py312-xdist
+        - macos: py312-xdist
           pytest-results-summary: true
-        - linux: py311-stdevdeps-xdist
-        - linux: py311-devdeps-xdist
-        - linux: py311-xdist-cov
+        - linux: py312-stdevdeps-xdist
```
[the remark this answers is not in the document]
Comment (hbushouse) [reply .github/workflows/ci.yml:76]: Given that we'll be delivering our next build under 3.11, I'd prefer to keep both the Linux and Macos entries using 3.11, but then use 3.12 for everything that involves latest/dev dependencies.

**Comment by tapastro** [comment]
Merging this should maybe have prompted a build - the readme is now updated to imply support of python 3.12, but no release supports 3.12 yet.

**Pull request merged**